libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthCoordinator.m

```diff
@@ -519,86 +535,76 @@ - (void)beginUserAgentFlow {
+    NSString *approvalUrlString = [self generateApprovalUrlString];
+    [self loadWebViewWithUrlString:approvalUrlString cookie:NO];
+}
 
-    // If an activation code is available (IP bypass flow), then use the "activated client" response type.
-    if (self.credentials.activationCode) {
-        [approvalUrl appendFormat:@"&%@=%@", kSFOAuthResponseType, kSFOAuthResponseTypeActivatedClientCode];
-    } else {
-        [approvalUrl appendFormat:@"&%@=%@", kSFOAuthResponseType, kSFOAuthResponseTypeToken];        
-    }
-        
-    NSString *scopeString = [self scopeQueryParamString];
-    if (scopeString != nil) {
-        [approvalUrl appendString:scopeString];
+- (void)beginJwtTokenExchangeFlow {
+    
+    if (![NSThread isMainThread]) {
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [self beginJwtTokenExchangeFlow];
+        });
+        return;
     }
     
-    // JWT Flow
-    if (self.credentials.jwt) {
-        [self swapJWTWithcompletionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
-            if (!error) {
-                bool swapOK = NO;
-                NSError *jsonError = nil;
-                id json = nil;
-                
-                json = [NSJSONSerialization JSONObjectWithData:data options:0 error:&jsonError];
-                if (nil == jsonError && [json isKindOfClass:[NSDictionary class]]) {
-                    NSDictionary *dict = (NSDictionary *)json;
-                    [self updateCredentials:dict];
-                    if (self.credentials.accessToken && self.credentials.apiUrl) {
-                        NSString *baseUrlString = [self.credentials.apiUrl absoluteString];
-                        NSString *escapedString = [approvalUrl stringByURLEncoding];
-                        NSString* approvalUrl = [NSString stringWithFormat:@"%@/secur/frontdoor.jsp?sid=%@&retURL=%@", baseUrlString, self.credentials.accessToken, escapedString];
-                        [self doLoadURL:approvalUrl withCookie:YES];
-                        swapOK = YES;
-                        self.credentials.jwt = nil;
-                    }
-                }
-                if (!swapOK) {
-                    NSError *error = [[self class] errorWithType:kSFOAuthErrorTypeJWTLaunchFailed description:@"The breeze link failed to launch."];
-                    [self notifyDelegateOfFailure:error authInfo:self.authInfo];
-                    self.credentials.jwt = nil;
-                }
-            }
-            else {
-                [self log:SFLogLevelError msg:[NSString stringWithFormat:@"Fail to swap JWT for access token: %@", [error localizedDescription]]];
+    NSAssert(self.credentials.jwt.length > 0, @"JWT token should be present at this point.");
+    
+    [self swapJWTWithCompletionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
+        if (error != nil) {
```

Comment:
The biggest difference here from the old completion handler code is just simplifying the conditionals, such that there isn't a lot of nested conditionals to look through when perusing the code.